File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        # base case. not a node
        if not root:
            return 0
        # base case. no children
        if not root.left and not root.right:
            return 1
        depth1, depth2 = 1, 1
        if root.left:
            depth1 += self.maxDepth(root.left)
        if root.right:
            depth2 += self.maxDepth(root.right)
        if depth1 >= depth2:
            return depth1
        else:
            return depth2
        

File: Data Structures & Algorithms/is-palindrome/submission-0.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        if (s.empty())
            return true;

        int i = 0;
        int j = s.size()-1;
        char* left;
        char* right;

        while (i < j) {
            left = &s[i];
            right = &s[j];


            while (
                ((*left < 48 || *left > 57) &&
                (*left < 65 || *left > 90) &&
                (*left < 97 || *left > 122)) && 
                (i < j)
                ) {
                    i++;
                    left = &s[i];
                }
            while (
                ((*right < 48 || *right > 57) &&
                (*right < 65 || *right > 90) &&
                (*right < 97 || *right > 122)) && 
                (i < j)
                ) {
                    j--;
                    right = &s[j];
                }
            // cout << *left << *right << endl; 
            if (tolower(*left) == tolower(*right)) {
                i++;
                j--;
            }
            else {
                return false;
            }     
        }
        return true;
    }
};
